src/board/BoardLogic.java

```diff
@@ -289,23 +297,66 @@ protected boolean removeApple(int x, int y) {
+        List<Field> validFields = new ArrayList<>(getValidAppleFields());
 
         // add apples until value of MAX_APPLES_ON_BOARD is reached
-        while (apples.size() < MAX_APPLES_ON_BOARD) {
-
-            // get random field with field status empty
-            Field appleField;
-            do {
-                appleField = getRandomField();
-
-            } while (!(fields[appleField.getPosX()][appleField.getPosY()].getState() == FieldState.Empty));
+        while (apples.size() < MAX_APPLES_ON_BOARD && validFields.size() > 0) {
+            // get random valid field
+            Field appleField = validFields.get((int) (Math.random() * validFields.size()));
+            validFields.remove(appleField);
 
             // set field state of random field to value apple and add apple to apple-list
-            fields[appleField.getPosX()][appleField.getPosY()].setState(FieldState.Apple);
+            appleField.setState(FieldState.Apple);
             apples.add(appleField);
         }
     }
 
+    /**
+     * generate a set of all valid apple fields
+     *
+     * @return a set containing all fields where an apple could be reached by a snake
+     */
+    protected Set<Field> getValidAppleFields() {
+        // checking from all snake heads
+        Deque<Field> checkFields = new ArrayDeque<>();
+        for (LinkedList<Field> snake : snakesLocation) {
+            checkFields.add(snake.getLast());
+        }
+
+        Set<Field> validFields = new HashSet<Field>();
+        while (!checkFields.isEmpty()) {
+            Field field = checkFields.removeLast();
+            Field tempField;
+            // checking in all 4 directions
+            if (field.getPosY() > 0
+                    && !validFields.contains(tempField = getFields()[field.getPosX()][field.getPosY() - 1])
+                    && tempField.getState() == FieldState.Empty) {
+                validFields.add(tempField);
+                checkFields.add(tempField);
+            }
+            if (field.getPosY() < SIZE_Y - 1
+                    && !validFields.contains(tempField = getFields()[field.getPosX()][field.getPosY() + 1])
+                    && tempField.getState() == FieldState.Empty) {
+                validFields.add(tempField);
+                checkFields.add(tempField);
+            }
+            if (field.getPosX() > 0
+                    && !validFields.contains(tempField = getFields()[field.getPosX() - 1][field.getPosY()])
+                    && tempField.getState() == FieldState.Empty) {
+                validFields.add(tempField);
+                checkFields.add(tempField);
+            }
+            if (field.getPosX() < SIZE_X - 1
+                    && !validFields.contains(tempField = getFields()[field.getPosX() + 1][field.getPosY()])
+                    && tempField.getState() == FieldState.Empty) {
+                validFields.add(tempField);
+                checkFields.add(tempField);
+            }
+        }
```

Comment:
This code performs 4 checks on every pixel, this can be reduced to 2 checks by using some sort of scan line algorithm e.g. the code I send you. (correct me if I'm wrong)